File: src/main/groovy/com/synopsys/integration/detect/workflow/project/DetectorNameVersionDecider.java
/**
 * synopsys-detect
 *
 * Copyright (c) 2019 Synopsys, Inc.
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package com.synopsys.integration.detect.workflow.project;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.synopsys.integration.detector.base.DetectorType;
import com.synopsys.integration.detect.workflow.project.decisions.ArbitraryNameVersionDecision;
import com.synopsys.integration.detect.workflow.project.decisions.NameVersionDecision;
import com.synopsys.integration.detect.workflow.project.decisions.PreferredDetectorDecision;
import com.synopsys.integration.detect.workflow.project.decisions.PreferredDetectorNotFoundDecision;
import com.synopsys.integration.detect.workflow.project.decisions.TooManyPreferredDetectorTypesFoundDecision;
import com.synopsys.integration.detect.workflow.project.decisions.UniqueDetectorDecision;
import com.synopsys.integration.detect.workflow.project.decisions.UniqueDetectorNotFoundDecision;
import com.synopsys.integration.util.NameVersion;

public class DetectorNameVersionDecider {
    private final Logger logger = LoggerFactory.getLogger(DetectorNameVersionDecider.class);

    public Optional<NameVersion> decideProjectNameVersion(final List<DetectorProjectInfo> projectNamePossibilities, final DetectorType preferredBomToolType) {
        final NameVersionDecision nameVersionDecision = decideProjectNameVersionFromDetector(projectNamePossibilities, preferredBomToolType);
        nameVersionDecision.printDescription(logger);
        return nameVersionDecision.getChosenNameVersion();
    }

    private NameVersionDecision decideProjectNameVersionFromDetector(final List<DetectorProjectInfo> projectNamePossibilities, final DetectorType preferredBomToolType) {
        final NameVersionDecision decision;

        if (preferredBomToolType != null) {
            final List<DetectorProjectInfo> possiblePreferred = projectNamePossibilities.stream()
                                                                    .filter(it -> it.getDetectorType() == preferredBomToolType)
                                                                    .collect(Collectors.toList());

            final List<DetectorProjectInfo> lowestDepthPossibilities = projectNamesAtLowestDepth(possiblePreferred);

            if (lowestDepthPossibilities.size() == 0) {
                decision = new PreferredDetectorNotFoundDecision(preferredBomToolType);
            } else if (lowestDepthPossibilities.size() == 1) {
                decision = new PreferredDetectorDecision(lowestDepthPossibilities.get(0));
            } else {
                decision = new TooManyPreferredDetectorTypesFoundDecision(preferredBomToolType);
            }
        } else {
            final List<DetectorProjectInfo> lowestDepthPossibilities = projectNamesAtLowestDepth(projectNamePossibilities);

            final Map<DetectorType, Long> lowestDepthTypeCounts = lowestDepthPossibilities.stream()
                                                                      .collect(Collectors.groupingBy(it -> it.getDetectorType(), Collectors.counting()));

            final List<DetectorType> singleInstanceLowestDepthBomTools = lowestDepthTypeCounts.entrySet().stream()
                                                                             .filter(it -> it.getValue() == 1)
                                                                             .map(it -> it.getKey())
                                                                             .collect(Collectors.toList());

            if (singleInstanceLowestDepthBomTools.size() == 1) {
                final DetectorType type = singleInstanceLowestDepthBomTools.get(0);
                final Optional<DetectorProjectInfo> chosen = lowestDepthPossibilities.stream().filter(it -> it.getDetectorType() == type).findFirst();

                if (chosen.isPresent()) {
                    decision = new UniqueDetectorDecision(chosen.get());
                } else {
                    decision = new UniqueDetectorNotFoundDecision();
                }
            } else if (singleInstanceLowestDepthBomTools.size() > 1) {
                decision = decideProjectNameVersionArbitrarily(lowestDepthPossibilities, singleInstanceLowestDepthBomTools);
            } else {
                decision = new UniqueDetectorNotFoundDecision();
            }
        }

        return decision;
    }

    private NameVersionDecision decideProjectNameVersionArbitrarily(final List<DetectorProjectInfo> possibilities, final List<DetectorType> bomToolOptions) {
        final List<DetectorProjectInfo> arbitraryOptions = possibilities.stream()
                                                               .filter(it -> bomToolOptions.contains(it.getDetectorType()))
                                                               .collect(Collectors.toList());

        final Optional<DetectorProjectInfo> chosen = arbitraryOptions.stream()
                                                         .sorted((o1, o2) -> o1.getNameVersion().getName().compareTo(o2.getNameVersion().getName()))
                                                         .findFirst();

        if (chosen.isPresent()) {
            return new ArbitraryNameVersionDecision(chosen.get(), arbitraryOptions);
        } else {
            return new UniqueDetectorNotFoundDecision();
        }
    }

    private List<DetectorProjectInfo> projectNamesAtLowestDepth(final List<DetectorProjectInfo> projectNamePossibilities) {
        List<DetectorProjectInfo> lowestDepthPossibilities = new ArrayList<>();
        final Optional<DetectorProjectInfo> bomToolProjectInfoAtLowestDepth = projectNamePossibilities.stream().min(Comparator.comparingInt(DetectorProjectInfo::getDepth));

        if (bomToolProjectInfoAtLowestDepth.isPresent()) {
            lowestDepthPossibilities = projectNamePossibilities.stream()
                                           .filter(it -> it.getDepth() == bomToolProjectInfoAtLowestDepth.get().getDepth())
                                           .collect(Collectors.toList());
        }

        return lowestDepthPossibilities;
    }

}

File: detect-kt/src/main/kotlin/DetectorNameVersionDecider.kt
package com.synopsys.integration.detect.kotlin.project

import com.synopsys.integration.detector.base.DetectorType
import com.synopsys.integration.util.NameVersion
import org.slf4j.LoggerFactory
import java.util.*

class DetectorNameVersionDecider {
    private val logger = LoggerFactory.getLogger(DetectorNameVersionDecider::class.java)

    fun decideProjectNameVersion(projectNamePossibilities: List<DetectorProjectInfo>, preferredBomToolType: DetectorType?): Optional<NameVersion> {
        val nameVersionDecision = decideProjectNameVersionFromDetector(projectNamePossibilities, preferredBomToolType)
        nameVersionDecision.printDescription(logger)
        return nameVersionDecision.chosenNameVersion
    }

    private fun decideProjectNameVersionFromDetector(projectNamePossibilities: List<DetectorProjectInfo>, preferredDetectorType: DetectorType?): NameVersionDecision {
        if (preferredDetectorType != null) {
            val preferredPossibilities = projectNamePossibilities.filter { it.detectorType == preferredDetectorType }
            val lowestDepthPossibilities = projectNamesAtLowestDepth(preferredPossibilities)
            val uniqueDetectorsAtLowestDepth = filterUniqueDetectorsOnly(lowestDepthPossibilities)

            return when {
                uniqueDetectorsAtLowestDepth.isEmpty() -> PreferredDetectorNotFoundDecision(preferredDetectorType)
                uniqueDetectorsAtLowestDepth.size == 1 -> PreferredDetectorDecision(uniqueDetectorsAtLowestDepth.first())
                else -> TooManyPreferredDetectorTypesFoundDecision(preferredDetectorType)
            }
        } else {
            val lowestDepthPossibilities = projectNamesAtLowestDepth(projectNamePossibilities)
            val uniqueDetectorsAtLowestDepth = filterUniqueDetectorsOnly(lowestDepthPossibilities)

            return when {
                uniqueDetectorsAtLowestDepth.size == 1 -> UniqueDetectorDecision(uniqueDetectorsAtLowestDepth.first())
                uniqueDetectorsAtLowestDepth.size > 1 -> decideProjectNameVersionArbitrarily(lowestDepthPossibilities)
                else -> UniqueDetectorNotFoundDecision()
            }
        }
    }

    private fun decideProjectNameVersionArbitrarily(arbitraryPossibilities: List<DetectorProjectInfo>): NameVersionDecision {
        val notGit = arbitraryPossibilities.filter { it.detectorType != DetectorType.GIT }

        val actualOptions = when {
            notGit.isEmpty() -> arbitraryPossibilities
            else -> notGit
        }

        val chosen = actualOptions.minBy { it.nameVersion.name }
        chosen?.let {
            val otherOptions = actualOptions.filter { it.detectorType != chosen.detectorType }
            return ArbitraryNameVersionDecision(Optional.of(chosen.nameVersion), chosen, otherOptions.toList())
        }

        return UniqueDetectorNotFoundDecision()
    }

    //Return only project info whose detector types appear exactly once.
    private fun filterUniqueDetectorsOnly(projectNamePossibilities: List<DetectorProjectInfo>): List<DetectorProjectInfo> {
        val grouped = projectNamePossibilities.groupBy { it.detectorType }
        val uniqueGroups = grouped.filter { it.value.size == 1 }
        return uniqueGroups.flatMap { it.value }
    }

    private fun projectNamesAtLowestDepth(projectNamePossibilities: List<DetectorProjectInfo>): List<DetectorProjectInfo> {
        val lowestDepth = projectNamePossibilities.minBy { it.depth }

        lowestDepth?.let {
            val allLowest = projectNamePossibilities.filter { it.depth == lowestDepth.depth }
            return allLowest.toList()
        }

        return emptyList()
    }
}
